message_ui: Show non-sticky recipient bar action buttons on hover.

ecxtacy opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #26852 

[CZO thread (The design idea discussion)](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20recipient.20bar/near/1598308)
[CZO thread (Implementation approach discussion)](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Show.20action.20buttons.20on.20hover.20.28recipient.20bars.29)

- Recipient bars show the action buttons on hover, except for the sticky recipient bar which always shows the action buttons.
- The CSS rule `visibility: hidden;` is toggled on hover, and `visibility: visible;` applied if the bar is sticky.
- The disappearance of the action buttons when action UI (notification menu, topic editing UI) is shown is prevented by using `:focus-within` CSS pseudo class on the recipient_bar to add `visibility: visible` to the action buttons.
- In case of conversation views, since there is a single message box present, the controls are made forever visible by using the CSS `:only-child` pseudo selector.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

### Final Look
![zulip-268](https://github.com/zulip/zulip/assets/11803841/269af734-bcca-49d5-831f-9bc993f32c9b)

<details><summary>Action UI remains intact</summary>
<img width="600" src="https://github.com/zulip/zulip/assets/11803841/e305502d-3bb1-4f45-9d34-7fedac930ec0">
</details>

<details>
<summary><strong>
Previously (All controls visible)
</strong></summary>
<img width="600" src="https://github.com/zulip/zulip/assets/11803841/fea24578-27a3-42b9-adef-221b65b48949">
</details>

<details><summary>Conversation view</summary>
<img src="https://github.com/zulip/zulip/assets/11803841/2b94919f-774b-4d08-8788-e44881a747c7" width="1000"/>
</details>


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks!

If a user is editing a topic, the editing UI should always be shown. It should not disappear if the mouse is moved.

<img width="552" alt="Screenshot 2023-10-18 at 11 50 35 AM" src="https://github.com/zulip/zulip/assets/2090066/441904e0-99b6-4966-a609-c90020f2a7e9">

alya commented:
Similarly, the notification settings button should remain visible while the menu is open, avoiding situations like this:

<img width="572" alt="Screenshot 2023-10-18 at 11 51 50 AM" src="https://github.com/zulip/zulip/assets/2090066/1f305dfe-4755-4a48-9bde-25b6ad709bb6">

ecxtacy commented:
Thanks a lot @alya, for bringing up these flaws, I will work on fixing them. :smiley:

ecxtacy commented:
@alya I have fixed the flaws which were produced and rebased + updated my branch. :smile: 

This is the UI after the fix :-
<img width="600" src="https://github.com/zulip/zulip/assets/11803841/e305502d-3bb1-4f45-9d34-7fedac930ec0">

alya commented:
The changes in response to the comments above look good! Please also address the comments [here](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20recipient.20bar/near/1665178) -- the icons should always be shown in a view that just has one conversation.

alya commented:
Thanks! Please post a screen capture or screenshot demonstrating the changes.

ecxtacy commented:
After the changes :-

<img src="https://github.com/zulip/zulip/assets/11803841/74c00227-63d2-4311-bf95-c2fbb3997511" width="700"/>

alya commented:
Works for me in manual testing! @amanagr are you up for reviewing this one?

amanagr commented:
LGTM!

timabbott commented on web/styles/zulip.css:
```diff
@@ -1881,6 +1898,13 @@ div.focused-message-list {
+    & .recipient_bar_controls {
+        visibility: visible;
+    }
+}
```
This check, unfortunately, will be true even if it's not a conversation view, but, say, a "mentions" or "starred" view that happens to only contain one message.

I think you'll need to key this fact off a fact about the view written into the `recipient_row.hbs` template, or maybe applied to `zfilt`. `is-conversation-view` would be a good name for a property added to that. You can add a new first commit that sets this.

And then this could be `div.focused-message-list.is-conversation-view ...`

ecxtacy replied:
Yes that is true, thank you for spotting this out ! :smile: 

As far as I've understood, I'll have to add some JS logic to check if the view is actually a **_conversation view_**, and if so, then I'll hand over this fact to the `recipient_row.hbs` template and then apply the approach you described.

ecxtacy commented:
I've updated the code, which addresses the bug mentioned by Tim. Now the action buttons are always visible only in conversation views, not any other view which might have a single message.

alya commented:
I did another round of manual testing and didn't spot any issues.

timabbott commented on web/src/message_list_view.js:
```diff
@@ -1003,6 +1003,13 @@ export class MessageListView {
+            $("#zfilt").addClass("is-conversation-view");
+        } else {
+            $("#zfilt").removeClass("is-conversation-view");
+        }
```
This is wrong if we are rerendering `message_lists.home`, which should not mutate `zfilt`. Please look at the `table_name` attribute for what part of the DOM to interact with.

ecxtacy replied:
Ill make the adjustments and update the code. Thank you for your feedback on this :smile:

timabbott commented:
Test-deploying this on chat.zulip.org.

ecxtacy commented:
Thanks ! :smile:

timabbott commented:
Can you rebase this and then ping me to deploy it in DMs? It needs a bit of a rework because `zfilt` no longer exists.

ecxtacy commented:
Sure :smile:

ecxtacy commented:
@timabbott I've rebased it.

timabbott commented:
@ecxtacy `zfilt` no longer exists in the app, so this code can't possibly work - can you please carefully update that and re-test this logic after doing so? It's important to always test your changes after a big rebase.

ecxtacy commented:
Yeah sure, I'll make the suitable changes on this. :+1: 
Thanks for pointing this out. :smile:

ecxtacy commented:
I have made the required changes. Thanks ! :smile:

AMANVISHWAKARMA27 commented:
Hi @ecxtacy, can I proceed with your PR ?

ecxtacy commented:
Yes sure 👍

timabbott commented:
Closing in favor of #32233 which continues this effort.
